lib/rust/ensogl/core/src/display/object/instance.rs

```diff
@@ -2024,3 +3616,1126 @@ mod tests {
+        let test = TestFlatChildren1::new();
+        test.root.set_size((10.0, 10.0));
+        test.node1.allow_grow().unsafe_set_forced_origin_alignment(alignment::Dim2::center());
+        test.run()
+            .assert_root_computed_size(10.0, 10.0)
+            .assert_node1_computed_size(10.0, 10.0)
+            .assert_root_position(0.0, 0.0)
+            .assert_node1_position(0.0, 0.0)
+            .assert_node1_content_origin(-5.0, -5.0)
+            .assert_root_content_origin(-5.0, -5.0);
+    }
+
+    /// ```text
+    /// 
+    ///        ╭────── ▶ ◀ ────────────────╮    ╭────── ▶ ◀ ───────────────╮
+    ///        │ node1                     │    │ node2                    │
+    /// ╭──────┼─────────────────────────┬─┼────┼────────────────────────╮ │
+    /// │ root │                         ┆ │    │                        │ │
+    /// │  ╭───┼──────╮                  ┆ ▼╭───┼──────╮                 │ ▼
+    /// │  │ node1_1  │                  ┆ ▲│ node2_1  │                 │ ▲
+    /// │  │   │      │  ╭──────────╮    ┆ ││   │      │  ╭─────────╮    │ │
+    /// │  ╰───┼──────╯  │ node1_2  │    ┆ │╰───┼──────╯  │ node2_2 │    │ │
+    /// │      ╰─────────┼──────────┼────┼─╯    ╰─────────┼─────────┼────┼─╯
+    /// │    ╱           │          │    ┆    ╱           │         │    │  
+    /// │  ◎             ╰──────────╯    ┆  ◎             ╰─────────╯    │  
+    /// ╰────────────────────────────────┴───────────────────────────────╯
+    /// ```
+    #[test]
+    fn test_horizontal_layout_with_children_with_forced_origins() {
+        let test = TestFlatChildren2::new();
+
+        test.root.use_auto_layout();
+        test.node1.set_size((2.0, 2.0));
+        test.node2.set_size((2.0, 2.0));
+
+        let node1_1 = test.node1.new_child_named("node1_1");
+        let node2_1 = test.node2.new_child_named("node2_1");
+        node1_1.allow_grow().unsafe_set_forced_origin_alignment(alignment::Dim2::center());
+        node2_1.allow_grow().unsafe_set_forced_origin_alignment(alignment::Dim2::center());
+
+        test.run();
+
+        assert_eq!(node1_1.position().xy(), Vector2(0.0, 0.0));
+        assert_eq!(node2_1.position().xy(), Vector2(0.0, 0.0));
+        assert_eq!(node1_1.computed_size(), Vector2(2.0, 2.0));
+        assert_eq!(node2_1.computed_size(), Vector2(2.0, 2.0));
+        assert_eq!(node1_1.content_origin(), Vector2(-1.0, -1.0));
+        assert_eq!(node2_1.content_origin(), Vector2(-1.0, -1.0));
+        assert_eq!(test.node1.content_origin(), Vector2(-1.0, -1.0));
+        assert_eq!(test.node2.content_origin(), Vector2(-1.0, -1.0));
+        assert_eq!(test.node1.computed_size(), Vector2(2.0, 2.0));
+        assert_eq!(test.node2.computed_size(), Vector2(2.0, 2.0));
+        assert_eq!(test.node1.position().xy(), Vector2(1.0, 1.0));
+        assert_eq!(test.node2.position().xy(), Vector2(3.0, 1.0));
+
+        test.assert_root_position(0.0, 0.0);
+    }
+}
```

Comment:
I think that 4K lines is a bit too many for a file. Can we move some categories (autolayout maybe?) to submodules?

Reply:
As agreed via phone, lets leave it as is because we do not want to expose huge amount of fns as pub crate.